consensus-peer/consensus/server/server.go

```diff
@@ -0,0 +1,186 @@
+	}
+	consensus.RegisterAtomicBroadcastServer(conn.Server, (*clientServer)(c))
+	return c
+}
+
+func (c *Server) RegisterConsenter(consensus fabric_consensus.Consenter) {
+	c.consensus = consensus
+	c.Executor = executor.NewImpl(c.consensus, c, c)
+	c.Executor.Start()
+}
+
+// gRPC atomic_broadcast interface
+func (c *clientServer) Broadcast(ctx context.Context, msg *consensus.Message) (*google_protobuf.Empty, error) {
+	// XXX check ctx tls credentials for permission to broadcast
+	// XXX log broadcast
+	now := time.Now()
+	c.consensus.RecvRequest(&pb.Transaction{
+		Payload: msg.Data,
+		Timestamp: &google_protobuf.Timestamp{
+			Seconds: now.Unix(),
+			Nanos:   int32(now.UnixNano() % 1000000000),
+		},
+	})
+	return &google_protobuf.Empty{}, nil
+}
+
+func (c *clientServer) BroadcastStream(srv consensus.AtomicBroadcast_BroadcastStreamServer) error {
+	for {
+		msg, err := srv.Recv()
+		if err != nil {
+			return err
+		}
+		now := time.Now()
+		c.consensus.RecvRequest(&pb.Transaction{
+			Payload: msg.Data,
+			Timestamp: &google_protobuf.Timestamp{
+				Seconds: now.Unix(),
+				Nanos:   int32(now.UnixNano() % 1000000000),
+			},
+		})
+		err = srv.Send(&consensus.Reply{})
+		if err != nil {
+			return err
+		}
+	}
+}
+
+func (c *clientServer) Deliver(_ *google_protobuf.Empty, srv consensus.AtomicBroadcast_DeliverServer) error {
+	// XXX check src tls credentials for permission to subscribe
+	// XXX log subscription
+	ch := make(chan *consensus.Block, c.queueSize)
+	c.lock.Lock()
+	c.deliver[ch] = struct{}{}
+	c.lock.Unlock()
+	for msg := range ch {
+		srv.Send(msg)
+	}
+	return nil
+}
+
+//
+func (c *Server) BeginTxBatch(id interface{}) error {
+	c.executed = &consensus.Block{}
+	return nil
+}
+
+func (c *Server) ExecTxs(id interface{}, txs []*pb.Transaction) ([]byte, error) {
+	msgs := make([]*consensus.Message, len(txs))
+	for i, tx := range txs {
+		msgs[i] = &consensus.Message{Data: tx.Payload}
+	}
+	c.executed.Messages = append(c.executed.Messages, msgs...)
+	return nil, nil
+}
+
+func (c *Server) CommitTxBatch(id interface{}, metadata []byte) (*pb.Block, error) {
+	b := c.executed
+	c.executed = &consensus.Block{}
+
+	c.lock.Lock()
+	out := c.deliver
+	c.lock.Unlock()
+
+	for ch := range out {
+		select {
+		case ch <- b:
+			break
+		default:
+			// ch is full, disconnect
+			c.lock.Lock()
+			delete(c.deliver, ch)
+			c.lock.Unlock()
+			close(ch)
+		}
+	}
+
+	return nil, nil
+}
+
+func (c *Server) RollbackTxBatch(id interface{}) error {
+	return nil
+}
+
+func (c *Server) PreviewCommitTxBatch(id interface{}, metadata []byte) ([]byte, error) {
+	panic("not implemented")
+}
+
+func (c *Server) GetBlockchainInfo() *pb.BlockchainInfo {
+	//	panic("not implemented")
+	return &pb.BlockchainInfo{}
+}
+
+// Syncer
+func (c *Server) Start() {
+}
+
+func (c *Server) Stop() {
+	panic("not implemented")
+}
+
+func (c *Server) SyncToTarget(blockNumber uint64, blockHash []byte, peerIDs []*pb.PeerID) (error, bool) {
+	panic("not implemented")
```

Comment:
I mean that in a PBFT network, only 2f+1 replicas are required to actually have any given state.  That if one of the replicas falls behind, and a client is subscribed to only it, you have a problem.

We could have the replica who falls behind (and realizes it) send a signal to the client to go find somebody else, but that seems more complicated to me than just to have the client subscribe to all streams, and enforce the 2f+1 commit message logic on the client side. This way, there's no complicated signalling and switching.

I still maintain eventually we'll want some lazy state transfer among the orderer nodes (assuming we keep some form of persistent log, limited or otherwise), as when a client is catching up, having to get failures back and try someone else seems suboptimal.